# Project Title:
Income Prediction Challenge For Azubian

# Business Understanding:

The "Income Prediction Challenge for Azubian" is a data-driven initiative that seeks to address the critical issue of income inequality in developing nations. The project focuses on utilizing machine learning techniques to predict whether an individual's income falls above or below a specific income threshold. By developing a robust predictive model, we aim to contribute to more accurate and cost-effective methods of monitoring key population indicators, such as income levels, between census years. This valuable information will empower policymakers to take more informed actions to mitigate and manage income inequality on a global scale.

## 1. Introduction:
This project aims to predict the estimated time of arrival (ETA) for Algerian ride-hailing business Yassir. ETA prediction is crucial in the ride-hailing industry to provide accurate and reliable arrival time estimates to passengers, thereby improving the user experience and operational efficiency.

### 1.1. Objectives:

- **Income Prediction Model:** 
The primary goal is to create a machine learning model capable of determining whether an individual's income exceeds a specified threshold.


- **Economic Inequality Mitigation:** 
By accurately predicting income levels, the project aims to support the reduction of income inequality by providing policymakers with critical insights.


- **Cost and Accuracy Improvement:** 
This solution endeavors to improve the efficiency of income-level monitoring by offering a more cost-effective and precise method compared to traditional census methods.

### 1.2. Methodology:

To achieve the project objectives, we will follow the Cross-Industry Standard Process for Data Mining (CRISP-DM) framework, a process used to guide the machine learning lifecycle. It is a six-phase process consisting of these key phases:

**i. Business Understanding:**
- Gain a deep understanding of the problem, its significance, and the potential impact of addressing income inequality.
- Define the objectives and the F1 score evaluation metric for model performance.

**ii. Data Understanding:**
- Load the provided training, testing and variable definitions datasets.
- Explore the provided datasets, including variable descriptions.
- Analyze the Train.csv dataset, which includes target income labels, to understand the data's structure and relationships.

**iii. Data Preparation:**
- Preprocess the data by handling missing values and data anomalies.
- Perform feature engineering to prepare the data for model training.

**iv. Modeling:**
- Select and implement machine learning algorithms suitable for classification tasks.
- Train the predictive models on the Train.csv dataset using features to predict income labels.

**v. Evaluation:**
- Assess the models' performance using the F1 score, which combines precision and recall, to ensure the model's accuracy and select the best performing model.
- Perform cross validation and hyperparameter tuning.

**vi. Deployment:**
- Deploy the trained model for prediction on the Test.csv dataset, which does not include target-related columns.
- The deployment to be done through one of these frameworks; Streamlit, Gradio or FastAPI through Docker Containerization and deployment on Hugging Face. This will allow external users to interact with the application.

By following the CRISP-DM framework, we aim to create a robust income prediction model that can effectively support efforts to address income inequality, provide policymakers with valuable insights, and improve the accuracy of population monitoring between census years. This project has the potential to make a significant impact on global economic equality.

# Data Understanding

## Setup

### Installations

### Importation of Relevant Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## Data Loading

### Loading the Variable Definitions, Train and Test Datasets

#### Variable Definitions Dataset

In [5]:
# Reading Variable Definitions dataset
def_df = pd.read_csv('VariableDefinitions.csv')
def_df

Column  \
0                              age   
1                           gender   
2                        education   
3                            class   
4              education_institute   
5                   marital_status   
6                             race   
7                      is_hispanic   
8            employment_commitment   
9              unemployment_reason   
10                 employment_stat   
11                   wage_per_hour   
12                  is_labor_union   
13           working_week_per_year   
14                   industry_code   
15              industry_code_main   
16                 occupation_code   
17            occupation_code_main   
18                  total_employed   
19                  household_stat   
20               household_summary   
21                 under_18_family   
22    veterans_admin_questionnaire   
23                     vet_benefit   
24                      tax_status   
25                           gains   
26                          losses   
27                   stocks_status   
28                     citizenship   
29                        mig_year   
30            country_of_birth_own   
31         country_of_birth_father   
32         country_of_birth_mother   
33    migration_code_change_in_msa   
34          migration_prev_sunbelt   
35  migration_code_move_within_reg   
36    migration_code_change_in_reg   
37            residence_1_year_ago   
38               old_residence_reg   
39             old_residence_state   
40            importance_of_record   
41              income_above_limit   

                                      Description  
0                               Age Of Individual  
1                                          Gender  
2                                       Education  
3                                 Class Of Worker  
4   Enrolled Educational Institution in last week  
5                                  Marital_Status  
6                                            Race  
7                                 Hispanic Origin  
8               Full Or Part Time Employment Stat  
9                         Reason For Unemployment  
10           Has Own Business Or Is Self Employed  
11                                  Wage Per Hour  
12                        Member Of A Labor Union  
13                         Weeks Worked In A Year  
14                                       Category  
15                            Major Industry Code  
16                                       Category  
17                          Major Occupation Code  
18                Num Persons Worked For Employer  
19             Detailed Household And Family Stat  
20        Detailed Household Summary In Household  
21                        Family Members Under 18  
22      Fill Inc Questionnaire For Veterans Admin  
23                                       Benefits  
24                               Tax Filer Status  
25                                          Gains  
26                                         Losses  
27                           Divdends From Stocks  
28                                    Citizenship  
29                                           Year  
30                     Individual's Birth Country  
31                         Father's Birth Country  
32                         Mother's Birth Country  
33                   Migration Code Change In Msa  
34                         Migration Prev Sunbelt  
35                 Migration Code Move Within Reg  
36                   Migration Code Change In Reg  
37                   Live In This House Year Ago?  
38                   Region Of Previous Residence  
39                    State Of Previous Residence  
40                             Weight Of Instance  
41                           Is income above 50k?

These variable definitions provide a clear understanding of the features and target variable used in the dataset, which is essential for data analysis and modeling.


#### Train Dataset

In [2]:
# Reading train dataset
train_df = pd.read_csv('Train.csv')
train_df.head()

ID  age   gender               education                class  \
0  ID_TZ0000   79   Female    High school graduate                  NaN   
1  ID_TZ0001   65   Female    High school graduate                  NaN   
2  ID_TZ0002   21     Male   12th grade no diploma   Federal government   
3  ID_TZ0003    2   Female                Children                  NaN   
4  ID_TZ0004   70     Male    High school graduate                  NaN   

  education_institute                    marital_status  \
0                 NaN                           Widowed   
1                 NaN                           Widowed   
2                 NaN                     Never married   
3                 NaN                     Never married   
4                 NaN   Married-civilian spouse present   

                         race is_hispanic      employment_commitment  ...  \
0                       White   All other         Not in labor force  ...   
1                       White   All other   Children or Armed Forces  ...   
2                       Black   All other   Children or Armed Forces  ...   
3   Asian or Pacific Islander   All other   Children or Armed Forces  ...   
4                       White   All other         Not in labor force  ...   

  country_of_birth_mother  migration_code_change_in_msa  \
0                      US                             ?   
1                      US                     unchanged   
2                      US                     unchanged   
3                   India                     unchanged   
4                      US                             ?   

   migration_prev_sunbelt migration_code_move_within_reg  \
0                       ?                              ?   
1                     NaN                      unchanged   
2                     NaN                      unchanged   
3                     NaN                      unchanged   
4                       ?                              ?   

   migration_code_change_in_reg  residence_1_year_ago old_residence_reg  \
0                             ?                   NaN               NaN   
1                     unchanged                  Same               NaN   
2                     unchanged                  Same               NaN   
3                     unchanged                  Same               NaN   
4                             ?                   NaN               NaN   

   old_residence_state importance_of_record  income_above_limit  
0                  NaN              1779.74         Below limit  
1                  NaN              2366.75         Below limit  
2                  NaN              1693.42         Below limit  
3                  NaN              1380.27         Below limit  
4                  NaN              1580.79         Below limit  

[5 rows x 43 columns]

#### Test Dataset

In [4]:
# Reading test dataset
test_df = pd.read_csv('Test.csv')
test_df.head()

ID  age   gender                    education     class  \
0  ID_TZ209499   54     Male         High school graduate   Private   
1  ID_TZ209500   53     Male             5th or 6th grade   Private   
2  ID_TZ209501   42     Male   Bachelors degree(BA AB BS)   Private   
3  ID_TZ209502   16   Female                    9th grade       NaN   
4  ID_TZ209503   16     Male                    9th grade       NaN   

  education_institute                    marital_status    race  \
0                 NaN   Married-civilian spouse present   White   
1                 NaN   Married-civilian spouse present   White   
2                 NaN   Married-civilian spouse present   White   
3         High school                     Never married   White   
4         High school                     Never married   White   

                  is_hispanic      employment_commitment  ...  \
0                   All other   Children or Armed Forces  ...   
1   Central or South American        Full-time schedules  ...   
2                   All other        Full-time schedules  ...   
3                   All other   Children or Armed Forces  ...   
4                   All other         Not in labor force  ...   

  country_of_birth_father  country_of_birth_mother  \
0                      US                       US   
1             El-Salvador              El-Salvador   
2                      US                       US   
3                      US                       US   
4                      US                       US   

   migration_code_change_in_msa migration_prev_sunbelt  \
0                     unchanged                    NaN   
1                             ?                      ?   
2                             ?                      ?   
3                     unchanged                    NaN   
4                             ?                      ?   

   migration_code_move_within_reg  migration_code_change_in_reg  \
0                       unchanged                     unchanged   
1                               ?                             ?   
2                               ?                             ?   
3                       unchanged                     unchanged   
4                               ?                             ?   

  residence_1_year_ago  old_residence_reg old_residence_state  \
0                 Same                NaN                 NaN   
1                  NaN                NaN                 NaN   
2                  NaN                NaN                 NaN   
3                 Same                NaN                 NaN   
4                  NaN                NaN                 NaN   

   importance_of_record  
0               3388.96  
1               1177.55  
2               4898.55  
3               1391.44  
4               1933.18  

[5 rows x 42 columns]

## Hypothesis    

**Null Hypothesis (H0):** There is no significant association between an individual's education level and the likelihood of having an income above the specified threshold.


**Alternative Hypothesis (H1):** Individuals with higher education levels are significantly more likely to have incomes above the specified threshold.


- The null hypothesis (H0) suggests that education level and income level are not related, meaning that having a higher education level does not increase the likelihood of earning an income above the threshold.

- The alternative hypothesis (H1) posits that there is a significant association between education level and income level, indicating that higher education levels are linked to a higher likelihood of having an income above the threshold.

We will conduct statistical tests to either accept or reject the null hypothesis in favor of the alternative hypothesis based on the evidence provided by the dataset.

## Key Analytical Questions:

To gain insights into the dataset and validate the hypothesis, we can formulate several key analytical questions for EDA:

i. What is the distribution of income levels in the dataset (above the threshold vs. below the threshold)?

- This question will provide an initial understanding of the balance between the two income categories.

ii. How does age relate to income levels in the dataset?

- Analyzing the age distribution among individuals with different income levels may reveal patterns related to age and income.

iii. Is there a significant gender-based income disparity?

- Exploring income levels based on gender can help us understand if gender plays a role in income categorization.

iv. What is the distribution of education levels among individuals with different income levels?

- Analyzing the educational backgrounds of individuals in both income categories can help assess the hypothesis regarding education and income.

v. Are there differences in employment status between the two income groups?

- Investigating the employment status and commitment of individuals based on income categories can provide insights into the relationship between employment and income.

vi. How do race and ethnicity correlate with income levels in the dataset?

- Understanding the distribution of income levels across different racial and ethnic groups can shed light on potential disparities.

vii. Are capital gains and losses associated with higher incomes?

- Examining the presence and amounts of capital gains and losses can help determine their impact on income levels.

viii. What is the relationship between occupation and income categories?

- Analyzing the occupation categories and their distribution among income groups can provide insights into employment roles.

ix. Do migration patterns or changes in residence relate to income levels?

- Investigating migration and residence changes may reveal how geographic mobility affects income.

x. How does tax status correspond to income levels?

Analyzing tax filing status can provide information about the impact of tax-related factors on income.

By exploring these key analytical questions, we can perform an in-depth EDA to uncover patterns, correlations, and potential factors influencing income levels. Additionally, it will help us determine whether our initial hypothesis regarding education and income holds true in this dataset.

## Exploratory Data Analysis (EDA):

### Understanding the datasets

An in-depth exploration of the datasets is presented to gain insights into the available variables,their distributions and relationships. This step will provide an initial undertanding of the datasets to identify any data quality issues that will inform the cleaning and pre-processing.

#### i. Column Information of The Datasets

In [3]:
# Column information of the train dataset
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 209499 entries, 0 to 209498
Data columns (total 43 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   ID                              209499 non-null  object 
 1   age                             209499 non-null  int64  
 2   gender                          209499 non-null  object 
 3   education                       209499 non-null  object 
 4   class                           104254 non-null  object 
 5   education_institute             13302 non-null   object 
 6   marital_status                  209499 non-null  object 
 7   race                            209499 non-null  object 
 8   is_hispanic                     209499 non-null  object 
 9   employment_commitment           209499 non-null  object 
 10  unemployment_reason             6520 non-null    object 
 11  employment_stat                 209499 non-null  int64  
 12  wage_per_hour   

- The train dataset contains 209,499 rows and 43 columns.


- The columns include various demographic and socioeconomic features, as well as the target variable "income_above_limit," which indicates whether the individual's income is above or below the income threshold ($50,000)


- Many columns contain missing values. For example, 'class,' 'education_institute,' 'unemployment_reason,' 'is_labor_union,' 'under_18_family,' 'veterans_admin_questionnaire,' 'old_residence_reg,' 'old_residence_state,' 'migration_code_change_in_msa,' 'migration_prev_sunbelt,' 'migration_code_move_within_reg,' and 'migration_code_change_in_reg' have significant missing values.

In [23]:
# Column information of the test dataset
test_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 89786 entries, 0 to 89785
Data columns (total 42 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   ID                              89786 non-null  object 
 1   age                             89786 non-null  int64  
 2   gender                          89786 non-null  object 
 3   education                       89786 non-null  object 
 4   class                           44707 non-null  object 
 5   education_institute             5616 non-null   object 
 6   marital_status                  89786 non-null  object 
 7   race                            89786 non-null  object 
 8   is_hispanic                     89786 non-null  object 
 9   employment_commitment           89786 non-null  object 
 10  unemployment_reason             2680 non-null   object 
 11  employment_stat                 89786 non-null  int64  
 12  wage_per_hour                   

#### ii. Shape of The Datasets

In [24]:
# The shape of the train dataset
train_df.shape

(209499, 43)

In [25]:
# The shape of the test dataset
test_df.shape

(89786, 42)

#### iii. Summary Statistics Datasets

In [26]:
# Summary Statistics of The Train Dataset
train_df.describe().round(3)

age  employment_stat  wage_per_hour  working_week_per_year  \
count  209499.000       209499.000     209499.000             209499.000   
mean       34.519            0.177         55.433                 23.159   
std        22.307            0.556        276.757                 24.398   
min         0.000            0.000          0.000                  0.000   
25%        15.000            0.000          0.000                  0.000   
50%        33.000            0.000          0.000                  8.000   
75%        50.000            0.000          0.000                 52.000   
max        90.000            2.000       9999.000                 52.000   

       industry_code  occupation_code  total_employed  vet_benefit  \
count     209499.000       209499.000      209499.000   209499.000   
mean          15.332           11.322           1.956        1.516   
std           18.050           14.461           2.365        0.851   
min            0.000            0.000           0.000        0.000   
25%            0.000            0.000           0.000        2.000   
50%            0.000            0.000           1.000        2.000   
75%           33.000           26.000           4.000        2.000   
max           51.000           46.000           6.000        2.000   

            gains      losses  stocks_status  mig_year  importance_of_record  
count  209499.000  209499.000     209499.000  209499.0            209499.000  
mean      435.927      36.882        194.533      94.5              1740.888  
std      4696.360     270.383       1956.376       0.5               995.560  
min         0.000       0.000          0.000      94.0                37.870  
25%         0.000       0.000          0.000      94.0              1061.290  
50%         0.000       0.000          0.000      94.0              1617.040  
75%         0.000       0.000          0.000      95.0              2185.480  
max     99999.000    4608.000      99999.000      95.0             18656.300

In [27]:
# Summary Statistics of The Test Dataset
test_df.describe().round(3)

age  employment_stat  wage_per_hour  working_week_per_year  \
count  89786.000        89786.000      89786.000              89786.000   
mean      34.586            0.176         54.339                 23.224   
std       22.346            0.554        265.198                 24.418   
min        0.000            0.000          0.000                  0.000   
25%       15.000            0.000          0.000                  0.000   
50%       33.000            0.000          0.000                  8.000   
75%       50.000            0.000          0.000                 52.000   
max       90.000            2.000       9400.000                 52.000   

       industry_code  occupation_code  total_employed  vet_benefit      gains  \
count      89786.000        89786.000       89786.000    89786.000  89786.000   
mean          15.377           11.298           1.956        1.518    421.978   
std           18.063           14.445           2.364        0.849   4610.516   
min            0.000            0.000           0.000        0.000      0.000   
25%            0.000            0.000           0.000        2.000      0.000   
50%            0.000            0.000           1.000        2.000      0.000   
75%           33.000           26.000           4.000        2.000      0.000   
max           51.000           46.000           6.000        2.000  99999.000   

          losses  stocks_status   mig_year  importance_of_record  
count  89786.000      89786.000  89786.000             89786.000  
mean      36.773        198.926     94.501              1738.264  
std      268.401       1893.917      0.500               990.837  
min        0.000          0.000     94.000                42.820  
25%        0.000          0.000     94.000              1059.115  
50%        0.000          0.000     95.000              1617.345  
75%        0.000          0.000     95.000              2193.735  
max     4608.000      99999.000     95.000             12960.200

#### iv. Checking for Missing Values in The Datasets

In [54]:
#define a function to print the missing values in the datasets
def show_missing_values(datasets):
    for name, data in datasets.items():
        title = f"Missing values in the {name.capitalize()} dataset:"
        underline = '*' * len(title)  # Create an underline of asterisks
        print(title)
        print(underline)
        
        # Calculate the sum of missing values for each column
        missing_values = data.isnull().sum()
        
        # Exclude columns with 0 as missing values
        missing_values = missing_values[missing_values > 0]
        
        # Check if there are any columns with missing values
        if not missing_values.empty:
            
            # Calculate the percentage of missing values
            missing_percentages = ((missing_values / len(data)) * 100).round(2)
            
            # Display the columns with missing values, the count of missing values, and their percentages
            missing_info = pd.DataFrame({
                'Column': missing_values.index, 
                'Missing Values': missing_values,
                'Missing Values Percentage (%)': missing_percentages
            })
            
            # Use to_string to remove index column
            print(missing_info.to_string(index=False))  
            print('===' * 26)
        else:
            # If no missing values found, indicate that
            print("No missing values found.")
        print()

# Call the function to show missing values in the datasets
show_missing_values(datasets)

Missing values in the Train dataset:
************************************
                        Column  Missing Values  Missing Values Percentage (%)
                         class          105245                          50.24
           education_institute          196197                          93.65
           unemployment_reason          202979                          96.89
                is_labor_union          189420                          90.42
          occupation_code_main          105694                          50.45
               under_18_family          151654                          72.39
  veterans_admin_questionnaire          207415                          99.01
  migration_code_change_in_msa            1588                           0.76
        migration_prev_sunbelt           88452                          42.22
migration_code_move_within_reg            1588                           0.76
  migration_code_change_in_reg            1588                      

In [22]:
# Loop through each column and print unique values
for column in train_df.columns:
    unique_values = train_df[column].unique()
    print(f"Column: {column}")
    print(f"Unique Values: {unique_values}")
    print('===' * 18)
    print()

Column: ID
Unique Values: ['ID_TZ0000' 'ID_TZ0001' 'ID_TZ0002' ... 'ID_TZ99997' 'ID_TZ99998'
 'ID_TZ99999']

Column: age
Unique Values: [79 65 21  2 70 45 53 22 73 30  4 16 43 36  5 88 40 47 59 69 50 27 39 85
 29 41 14 33 67 52 11  9 13 19 26 23 37 58 63 46 62 28 31  3 18 78 15 38
  7 35  1 20  0 48 24 56 25  8 66 71 32 75 51 10 44 42 55 74 77 34 80 17
 83 86 12 68 60 57 64 72 90 61 82  6 84 49 76 54 89 81 87]

Column: gender
Unique Values: [' Female' ' Male']

Column: education
Unique Values: [' High school graduate' ' 12th grade no diploma' ' Children'
 ' Bachelors degree(BA AB BS)' ' 7th and 8th grade' ' 11th grade'
 ' 9th grade' ' Masters degree(MA MS MEng MEd MSW MBA)' ' 10th grade'
 ' Associates degree-academic program' ' 1st 2nd 3rd or 4th grade'
 ' Some college but no degree' ' Less than 1st grade'
 ' Associates degree-occup /vocational'
 ' Prof school degree (MD DDS DVM LLB JD)' ' 5th or 6th grade'
 ' Doctorate degree(PhD EdD)']

Column: class
Unique Values: [nan ' Federal gov